src/protocol/prf_sharding/feature_label_dot_product.rs

```diff
@@ -0,0 +1,469 @@
+/// device can be processed together.
+///
+/// This circuit expects to receive records from multiple users,
+/// but with all of the records from a given user adjacent to one another, and in time order.
+///
+/// This circuit will compute attribution, and per-user capping.
+///
+/// The output of this circuit is the input to the next stage: Aggregation.
+///
+/// # Errors
+/// Propagates errors from multiplications
+/// # Panics
+/// Propagates errors from multiplications
+pub async fn compute_feature_label_dot_product<C, FV, F, S>(
+    sh_ctx: C,
+    input_rows: Vec<PrfShardedIpaInputRow<FV>>,
+) -> Result<Vec<S>, Error>
+where
+    C: UpgradableContext,
+    C::UpgradedContext<Gf2>: UpgradedContext<Gf2, Share = Replicated<Gf2>>,
+    C::UpgradedContext<F>: UpgradedContext<F, Share = S>,
+    S: LinearSecretSharing<F> + Serializable + SecureMul<C::UpgradedContext<F>>,
+    FV: GaloisField,
+    F: PrimeField + ExtendableField,
+{
+    assert!(FV::BITS > 0);
+
+    let rows_chunked_by_user = chunk_rows_by_user(input_rows);
+    let histogram = compute_histogram_of_users_with_row_count(&rows_chunked_by_user);
+    let binary_validator = sh_ctx.narrow(&Step::BinaryValidator).validator::<Gf2>();
+    let binary_m_ctx = binary_validator.context();
+    let mut num_users_who_encountered_row_depth = Vec::with_capacity(histogram.len());
+    let ctx_for_row_number = set_up_contexts(&binary_m_ctx, &histogram);
+    let mut futures = Vec::with_capacity(rows_chunked_by_user.len());
+    for rows_for_user in rows_chunked_by_user {
+        for i in 0..rows_for_user.len() {
+            if i >= num_users_who_encountered_row_depth.len() {
+                num_users_who_encountered_row_depth.push(0);
+            }
+            num_users_who_encountered_row_depth[i] += 1;
+        }
+
+        futures.push(evaluate_per_user_attribution_circuit(
+            &ctx_for_row_number,
+            num_users_who_encountered_row_depth
+                .iter()
+                .take(rows_for_user.len())
+                .map(|x| RecordId(x - 1))
+                .collect(),
+            rows_for_user,
+        ));
+    }
+    let outputs_chunked_by_user = seq_try_join_all(sh_ctx.active_work(), futures)
+        .await?
+        .into_iter()
+        .flatten()
+        .collect::<Vec<BitDecomposed<Replicated<Gf2>>>>();
+
+    let prime_field_validator = sh_ctx.narrow(&Step::PrimeFieldValidator).validator::<F>();
+    let prime_field_ctx = prime_field_validator.context();
+
+    // modulus convert feature vector bits
+    let converted_feature_vector_bits = convert_bits(
+        prime_field_ctx
+            .narrow(&Step::ModulusConvertFeatureVectorBits)
+            .set_total_records(outputs_chunked_by_user.len()),
+        stream_iter(outputs_chunked_by_user),
+        0..FV::BITS,
+    );
+
+    converted_feature_vector_bits
+        .try_fold(
+            vec![S::ZERO; 1 << FV::BITS],
```

Comment:
```suggestion
            vec![S::ZERO; usize::try_from(FV::BITS).unwrap()],
```

makes allocator's life a bit easier  :)

Reply:
Thank you Alex! Good find =).